Vue 3 Support

With Vue3 on the horizon, this library should probably support it. I've gotten started with Vue 3 in one of my projects, so I'll try and fork this and open a PR when I have time.

- `vue@3.0.0-rc.5`
- `vue-recaptcha-v3@1.9.0`

### Typescript

Looks like this library is missing an `install` method?

```ts
import { VueReCaptcha } from 'vue-recaptcha-v3';

createApp(App)
  .use(VueReCaptcha, { siteKey: '<token>' })
  .mount('#app');
```

```
Argument of type '(Vue: typeof import("<project>/node_modules/vue/dist/vue"), options: IReCaptchaOptions) => void' is not assignable to parameter of type 'Plugin_2'.
  Type '(Vue: typeof import("<project>/node_modules/vue/dist/vue"), options: IReCaptchaOptions) => void' is not assignable to type 'PluginInstallFunction & { install?: PluginInstallFunction | undefined; }'.
    Type '(Vue: typeof import("<project>/node_modules/vue/dist/vue"), options: IReCaptchaOptions) => void' is not assignable to type 'PluginInstallFunction'.
      Types of parameters 'Vue' and 'app' are incompatible.
        Type 'App<any>' is missing the following properties from type 'typeof import("node_modules/vue/dist/vue")': compile, useCssModule, useCssVars, createApp, and 104 more.
```

### Usage

If you ignore the TS error and continue on, you get an error at runtime because there is no longer a default export from `vue`:

```
Uncaught TypeError: Vue.prototype is undefined
    > VueReCaptcha ReCaptchaVuePlugin.js:46
```

Thanks. Any working example would be good. I am getting this in my package.json

"vue-recaptcha-v3": "file:local_modules/vue-recaptcha-v3-1.9.0.tgz",

and its saying failed to compile..

./src/main.js
Module not found: Error: Can't resolve 'vue-recaptcha-v3' in '/Applications/MAMP/htdocs/vue/myfoldername/src'

I already have an App component so not sure how to do 2 app components in main.js which I believe this code should go..

--- a/README.md
+++ b/README.md
@@ -43,6 +43,49 @@ new Vue({
 })
 ```
 
+### Vue 3 Composition
+
+```js
+import Vue from 'vue'
+import { VueReCaptcha, useRecaptcha } from 'vue-recaptcha-v3'
+
+// For more options see below
+Vue.use(VueReCaptcha, { siteKey: '<site key>', useComposition: true })
+
+new Vue({
+  setup() {
+    const { executeRecaptcha, recaptchaLoaded } = useRecaptcha();
+
+    const recaptcha = async () => {
+      // (optional) Wait until recaptcha has been loaded.
+      await recaptchaLoaded()
+
+      // Execute reCAPTCHA with action "login".
+      const token = await executeRecaptcha('login')
+    }
+
+    return {
+      recaptcha,
+    }
+  },
+  template: '<button @click="recaptcha">Execute recaptcha</button>'
+})
+```
+
+### TypeScript + Vue 3
+To get type suggestions for instance variables (this is not needed for composition API), create a new file called `vue-recaptcha-shims.d.ts` and put the following inside it:
+```ts
+import { ReCaptchaInstance } from 'recaptcha-v3';
+
+declare module '@vue/runtime-core' {
+  interface ComponentCustomProperties {
+    $recaptcha: (action: string) => Promise<string>;
+    $recaptchaLoaded: () => Promise<boolean>;
+    $recaptchaInstance: ReCaptchaInstance;
+  }
+}
+```
+
 ## Options
 This plugin offers optional options to configure the behavior of some parts.
 
@@ -80,4 +123,4 @@ recaptcha.hideBadge()
 
 // Show reCAPTCHA badge:
 recaptcha.showBadge()
-```  
+```

--- a/README.md
+++ b/README.md
@@ -75,7 +75,7 @@ createApp(component)
 ```
 
 ### TypeScript + Vue 3
-To get type suggestions for instance variables (this is not needed for composition API), create a new file called `vue-recaptcha-shims.d.ts` and put the following inside it:
+To get type suggestions for instance variables (this is not needed for composition API), create a new file called `shims-vue-recaptcha-v3.d.ts` and put the following inside it:
 ```ts
 import { ReCaptchaInstance } from 'recaptcha-v3'
 

--- a/README.md
+++ b/README.md
@@ -46,15 +46,12 @@ new Vue({
 ### Vue 3 Composition
 
 ```js
-import Vue from 'vue'
+import { createApp } from 'vue'
 import { VueReCaptcha, useRecaptcha } from 'vue-recaptcha-v3'
 
-// For more options see below
-Vue.use(VueReCaptcha, { siteKey: '<site key>', useComposition: true })
-
-new Vue({
+const component = {
   setup() {
-    const { executeRecaptcha, recaptchaLoaded } = useRecaptcha();
+    const { executeRecaptcha, recaptchaLoaded } = useRecaptcha()
 
     const recaptcha = async () => {
       // (optional) Wait until recaptcha has been loaded.
@@ -62,26 +59,31 @@ new Vue({
 
       // Execute reCAPTCHA with action "login".
       const token = await executeRecaptcha('login')
+
+      // Do stuff with the received token.
     }
 
     return {
-      recaptcha,
+      recaptcha
     }
   },
   template: '<button @click="recaptcha">Execute recaptcha</button>'
-})
+}
+
+createApp(component)
+  .use(VueReCaptcha, { siteKey: '<site key>' })
 ```
 
 ### TypeScript + Vue 3
 To get type suggestions for instance variables (this is not needed for composition API), create a new file called `vue-recaptcha-shims.d.ts` and put the following inside it:
 ```ts
-import { ReCaptchaInstance } from 'recaptcha-v3';
+import { ReCaptchaInstance } from 'recaptcha-v3'
 
 declare module '@vue/runtime-core' {
   interface ComponentCustomProperties {
-    $recaptcha: (action: string) => Promise<string>;
-    $recaptchaLoaded: () => Promise<boolean>;
-    $recaptchaInstance: ReCaptchaInstance;
+    $recaptcha: (action: string) => Promise<string>
+    $recaptchaLoaded: () => Promise<boolean>
+    $recaptchaInstance: ReCaptchaInstance
   }
 }
 ```

--- a/demo/src/index.ts
+++ b/demo/src/index.ts
@@ -14,8 +14,8 @@ const compoent = {
     }
   },
   template: '<button @click="recaptcha">Execute recaptcha</button>'
-};
+}
 
 createApp(compoent)
   .use(VueReCaptcha, { siteKey: '6LfC6HgUAAAAAEtG92bYRzwYkczElxq7WkCoG4Ob' })
-  .mount('#inject');
+  .mount('#inject')

--- a/demo/src/index.ts
+++ b/demo/src/index.ts
@@ -1,9 +1,7 @@
-import Vue from 'vue'
+import { createApp } from 'vue'
 import { VueReCaptcha } from '../../src/ReCaptchaVuePlugin'
 
-Vue.use(VueReCaptcha, { siteKey: '6LfC6HgUAAAAAEtG92bYRzwYkczElxq7WkCoG4Ob' })
-
-new Vue({
+const compoent = {
   methods: {
     recaptcha () {
       console.log('recaptcha clicked')
@@ -16,4 +14,8 @@ new Vue({
     }
   },
   template: '<button @click="recaptcha">Execute recaptcha</button>'
-}).$mount('#inject')
+};
+
+createApp(compoent)
+  .use(VueReCaptcha, { siteKey: '6LfC6HgUAAAAAEtG92bYRzwYkczElxq7WkCoG4Ob' })
+  .mount('#inject');

--- a/demo/tsconfig.json
+++ b/demo/tsconfig.json
@@ -8,7 +8,7 @@
     "declaration": true,
     "removeComments": true,
     "lib": [
-      "es6",
+      "esnext",
       "dom"
     ]
   },

--- a/package.json
+++ b/package.json
@@ -43,12 +43,12 @@
     "html-webpack-plugin": "^4.5.0",
     "ts-loader": "^8.0.4",
     "typescript": "^4.0.3",
-    "vue": "^2.6.12",
+    "vue": "^3.0.0",
     "webpack": "^4.44.2",
     "webpack-cli": "^3.3.12",
     "webpack-dev-server": "^3.11.0"
   },
   "peerDependencies": {
-    "vue": "^2.6.12"
+    "vue": "^3.0.0"
   }
 }

--- a/src/ReCaptchaVuePlugin.ts
+++ b/src/ReCaptchaVuePlugin.ts
@@ -21,8 +21,8 @@ export const VueReCaptcha = {
     app.config.globalProperties.$recaptchaLoaded = recaptchaLoaded(isLoaded)
 
     initializeReCaptcha(options).then((wrapper) => {
-      instance.value = wrapper
       isLoaded.value = true
+      instance.value = wrapper
 
       app.config.globalProperties.$recaptcha = recaptcha(instance)
       app.config.globalProperties.$recaptchaInstance = instance

--- a/src/ReCaptchaVuePlugin.ts
+++ b/src/ReCaptchaVuePlugin.ts
@@ -6,11 +6,11 @@ const VueReCaptchaInjectKey: InjectionKey<IReCaptchaComposition> = Symbol('VueRe
 
 interface IGlobalConfig {
   loadedWaiters: Array<({resolve: (resolve: boolean) => void, reject: (reject: Error) => void})>
-  error: Error | undefined
+  error: Error | null
 }
 const globalConfig: IGlobalConfig = {
   loadedWaiters: [],
-  error: undefined
+  error: null
 }
 
 export const VueReCaptcha = {

--- a/src/ReCaptchaVuePlugin.ts
+++ b/src/ReCaptchaVuePlugin.ts
@@ -5,14 +5,12 @@ import { IReCaptchaOptions } from './IReCaptchaOptions'
 const VueReCaptchaInjectKey: InjectionKey<IReCaptchaComposition> = Symbol('VueReCaptchaInjectKey')
 
 const globalConfig = {
-  options: undefined as IReCaptchaOptions,
   loadedWaiters: [] as Array<({resolve: (resolve: boolean) => void, reject: (reject: Error) => void})>,
   error: undefined as Error | undefined
 }
 
 export const VueReCaptcha = {
   install (app: App, options: IReCaptchaOptions) {
-    globalConfig.options = options
     const isLoaded = ref(false)
     const instance: Ref<ReCaptchaInstance | undefined> = ref(undefined)
 

--- a/src/ReCaptchaVuePlugin.ts
+++ b/src/ReCaptchaVuePlugin.ts
@@ -4,9 +4,13 @@ import { IReCaptchaOptions } from './IReCaptchaOptions'
 
 const VueReCaptchaInjectKey: InjectionKey<IReCaptchaComposition> = Symbol('VueReCaptchaInjectKey')
 
-const globalConfig = {
-  loadedWaiters: [] as Array<({resolve: (resolve: boolean) => void, reject: (reject: Error) => void})>,
-  error: undefined as Error | undefined
+interface IGlobalConfig {
+  loadedWaiters: Array<({resolve: (resolve: boolean) => void, reject: (reject: Error) => void})>
+  error: Error | undefined
+}
+const globalConfig: IGlobalConfig = {
+  loadedWaiters: [],
+  error: undefined
 }
 
 export const VueReCaptcha = {

--- a/src/ReCaptchaVuePlugin.ts
+++ b/src/ReCaptchaVuePlugin.ts
@@ -16,13 +16,13 @@ export const VueReCaptcha = {
     const isLoaded = ref(false)
     const instance: Ref<ReCaptchaInstance | undefined> = ref(undefined)
 
-    app.config.globalProperties.$recaptchaLoaded = recaptchaLoaded
+    app.config.globalProperties.$recaptchaLoaded = recaptchaLoaded(isLoaded)
 
     initializeReCaptcha(options).then((wrapper) => {
       instance.value = wrapper
       isLoaded.value = true
 
-      app.config.globalProperties.$recaptcha = recaptcha
+      app.config.globalProperties.$recaptcha = recaptcha(instance)
       app.config.globalProperties.$recaptchaInstance = instance
 
       globalConfig.loadedWaiters.forEach((v) => v.resolve(true))

--- a/src/ReCaptchaVuePlugin.ts
+++ b/src/ReCaptchaVuePlugin.ts
@@ -1,49 +1,75 @@
 import { load as loadReCaptcha, ReCaptchaInstance } from 'recaptcha-v3'
-import _Vue from 'vue'
+import { App, Ref, ref, provide, inject, InjectionKey } from 'vue'
 import { IReCaptchaOptions } from './IReCaptchaOptions'
 
-export function VueReCaptcha (Vue: typeof _Vue, options: IReCaptchaOptions): void {
-  const plugin = new ReCaptchaVuePlugin()
-  let recaptchaLoaded = false
-  let recaptchaError: Error = null
+const VueReCaptchaInjectKey: InjectionKey<IReCaptchaComposition> = Symbol('VueReCaptchaInjectKey')
 
-  const loadedWaiters: Array<({resolve: (resolve: boolean) => void, reject: (reject: Error) => void})> = []
+const globalConfig = {
+  options: undefined as IReCaptchaOptions,
+  loadedWaiters: [] as Array<({resolve: (resolve: boolean) => void, reject: (reject: Error) => void})>,
+  error: undefined as Error | undefined
+}
 
+export const VueReCaptcha = {
+  install (app: App, options: IReCaptchaOptions) {
+    globalConfig.options = options
+    const isLoaded = ref(false)
+    const instance: Ref<ReCaptchaInstance | undefined> = ref(undefined)
+
+    app.config.globalProperties.$recaptchaLoaded = recaptchaLoaded
+
+    initializeReCaptcha(options).then((wrapper) => {
+      instance.value = wrapper
+      isLoaded.value = true
+
+      app.config.globalProperties.$recaptcha = recaptcha
+      app.config.globalProperties.$recaptchaInstance = instance
+
+      globalConfig.loadedWaiters.forEach((v) => v.resolve(true))
+    }).catch((error) => {
+      globalConfig.error = error
+      globalConfig.loadedWaiters.forEach((v) => v.reject(error))
+    })
+
+    provide(VueReCaptchaInjectKey, {
+      instance,
+      isLoaded,
+      executeRecaptcha: recaptcha(instance),
+      recaptchaLoaded: recaptchaLoaded(isLoaded)
+    })
+  }
+}
+
+export function useReCaptcha (): IReCaptchaComposition {
+  return inject(VueReCaptchaInjectKey)
+}
+
+async function initializeReCaptcha (options: IReCaptchaOptions): Promise<ReCaptchaInstance> {
+  return await loadReCaptcha(options.siteKey, options.loaderOptions)
+}
+
+function recaptchaLoaded (isLoaded: Ref<boolean>) {
   // eslint-disable-next-line @typescript-eslint/promise-function-async
-  Vue.prototype.$recaptchaLoaded = () => new Promise<boolean>((resolve, reject) => {
-    if (recaptchaError !== null) {
-      return reject(recaptchaError)
+  return () => new Promise<boolean>((resolve, reject) => {
+    if (globalConfig.error !== null) {
+      return reject(globalConfig.error)
     }
-    if (recaptchaLoaded) {
+    if (isLoaded.value) {
       return resolve(true)
     }
-    loadedWaiters.push({ resolve, reject })
-  })
-
-  plugin.initializeReCaptcha(options).then((wrapper) => {
-    Vue.prototype.$recaptcha = async (action: string): Promise<string> => {
-      return await wrapper.execute(action)
-    }
-
-    Vue.prototype.$recaptchaInstance = wrapper
-    recaptchaLoaded = true
-    loadedWaiters.forEach((v) => v.resolve(true))
-  }).catch((error) => {
-    recaptchaError = error
-    loadedWaiters.forEach((v) => v.reject(error))
+    globalConfig.loadedWaiters.push({ resolve, reject })
   })
 }
 
-class ReCaptchaVuePlugin {
-  public async initializeReCaptcha (options: IReCaptchaOptions): Promise<ReCaptchaInstance> {
-    return await loadReCaptcha(options.siteKey, options.loaderOptions)
+function recaptcha (instance: Ref<ReCaptchaInstance | undefined>) {
+  return async (action: string): Promise<string> => {
+    return await instance.value?.execute(action)
   }
 }
 
-declare module 'vue/types/vue' {
-  interface Vue {
-    $recaptcha: (action: string) => Promise<string>
-    $recaptchaLoaded: () => Promise<boolean>
-    $recaptchaInstance: ReCaptchaInstance
-  }
+export interface IReCaptchaComposition {
+  isLoaded: Ref<boolean>
+  instance: Ref<ReCaptchaInstance | undefined>
+  executeRecaptcha: (action: string) => Promise<string>
+  recaptchaLoaded: () => Promise<boolean>
 }

--- a/src/ReCaptchaVuePlugin.ts
+++ b/src/ReCaptchaVuePlugin.ts
@@ -33,7 +33,7 @@ export const VueReCaptcha = {
       globalConfig.loadedWaiters.forEach((v) => v.reject(error))
     })
 
-    provide(VueReCaptchaInjectKey, {
+    app.provide(VueReCaptchaInjectKey, {
       instance,
       isLoaded,
       executeRecaptcha: recaptcha(instance),

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -8,7 +8,7 @@
     "declaration": true,
     "removeComments": true,
     "lib": [
-      "es6",
+      "esnext",
       "dom"
     ]
   },

